### ISSUE TITLE
The ability to insert and or append rows

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#556]
_Original author: steve.mo...@gmail.com (March 22, 2012 15:52:44)_

I have been using google refine a lot.  PS think it's awesome.

When cleaning up data regularly I also need to add in new rows of information into the data set.  Currently I have to export the data and then do the inserts in textmate or excel.

So just want the ability to insert rows and to manually edit the values in the cells.

_Original issue: http://code.google.com/p/google-refine/issues/detail?id=556_

**Comment by ostephens** [comment]
I've just released an extension that allows the addition of blank rows into an existing project. This is a step up from the work around described by @tfmorris above, and may meet the use case described in the original issue. However it doesn't go as far as #715

**Comment by olea** [comment]
> I've just released an extension[1] 

I guess this is not the best place to report but I've just tried the extension on OR 3.4-beta and got this error:

```
12:34:24.980 [                butterfly] Error loading special module manager (344ms)
java.lang.ClassNotFoundException: com.k_int.gokb.module.GOKbModuleImpl
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at edu.mit.simile.butterfly.Butterfly.createModule(Butterfly.java:671)
	at edu.mit.simile.butterfly.Butterfly.configure(Butterfly.java:412)
	at edu.mit.simile.butterfly.Butterfly.init(Butterfly.java:308)
	at org.mortbay.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:440)
	at org.mortbay.jetty.servlet.ServletHolder.doStart(ServletHolder.java:263)
	at com.google.refine.RefineServer.configure(Refine.java:291)
	at com.google.refine.RefineServer.init(Refine.java:203)
	at com.google.refine.Refine.init(Refine.java:109)
	at com.google.refine.Refine.main(Refine.java:103)
```

FYI

**Comment by ostephens** [comment]
I'm afraid that the extension has not been updated to work with the latest versions of OpenRefine. I haven't tested it across versions but it was originally written to work with v2.7, and my guess would be it will work OK up to v3.1 (but I've not tested it).

I migrated some of the functionality from the extension into the core product https://github.com/OpenRefine/OpenRefine/pull/1993 but this didn't include the "add blank rows" option as it didn't feel that it added enough value beyond the hack that Tom describes above:

> Add a cell value of ",,,,,,,,,,,,,,,,,,,,,,,,,,," (or something similar)
> Use the Edit Cells -> Split multivalued cells to create a bunch of empty rows

**Comment by olea** [comment]
@ostephens thanks :-)

**Comment by akellerhals** [comment]
Hi, I‘m fairly late in these thead but I tried that „Split multivalued cells to create an empty row“. Result: I have a additional row which, however, is depending from the ‚parent row‘ and has no own row number. Can I clearly separate this splitting based row from the original row?

**Comment by ostephens** [comment]
@akellerhals it sounds like you are working in "records" mode. You need to switch to "row" mode to see the individual row number - see https://docs.openrefine.org/manual/exploring/#rows-vs-records for more information

**Comment by akellerhals** [comment]
@ostephens Thanks, problem solved, looks like a detail but has escaped my attention, will read the links,

**Comment by strarsis** [comment]
Then there is the case where a cell is split in record mode, but the newly created riws end up in a new, separate record. Which shouldn't happen.

**Comment by ostephens** [comment]
@strarsis the way records work is very specific - it groups together rows based on the first column(s) being populated (or not) - it's explained more at https://docs.openrefine.org/manual/exploring/#rows-vs-records

This means that a Record isn't a static grouping of roads and the arrangement of rows into records can change as you change the content of cells

**Comment by strarsis** [comment]
@ostephens: Sometimes, when splitting cells a row is introduced that is wholly empty. This is interpreted by OpenRefine as a new, independent record. How can I get rid of the empty row? When the records are exported, interestingly, the resulting set doesn't contain these empty-row records, those rows belong then to their original records again.

**Comment by ostephens** [comment]
@strarsis I think the issue you are describing deserves it's own issue as this issue is a request for a feature but it sounds like you are describing a possible bug. Would you be willing to open a new issue with a description of the scenario that leads to the problem you are describing?

**Comment by magdmartin** [comment]
duplicate of #1855 ?

**Comment by thadguidry** [comment]
@magdmartin Yes, I think #1855 is a duplicate request for the same needs of the insert rows at X index as this issue.  Once someone finishes work on that #1855 (and @steve-kasica basically completed it there, just lacking tests) then I would say we can close this very old issue #556 as well.  

- [ ] Once that is done and closed, then Records mode can possibly be enhanced as well where Split Cells could offer the checkbox option to insert empty rows or not.

**Comment by tfmorris** [comment]
Yes, a duplicate. #1855 was a request for a specific implementation, which we're not actually going to use, so the PR might have been better placed here, but, just for convenience sake, let's close this issue even though it's older.
